### ISSUE TITLE
Does Quant-Noise with iterative Product Quantization actually use the Straight-Through-Estimator (STE)?

### ISSUE THREAD

**Issue by pdeubel** [pytorch/fairseq#3195]
### Before asking:   
1. search the issues.
- Maybe this #3163 is related
2. search the docs.
- Done

#### What is your question?

I am trying to understand the Quant-Noise approach with iterative Product Quantization (iPQ). For that I used a minimal example for a ConvNet from the PyTorch documentation to observe the method.

In [your documentation on Quant-Noise](https://github.com/pytorch/fairseq/tree/master/examples/quant_noise#iterative-product-quantization) you mention that first, the module needs to be trained uncompressed with Quant-Noise and then trained with iPQ.

First question: Does this mean that Quant-Noise uses the proxy noise function (i.e. zeroing) and after that the resulting network will be trained with iPQ where actual centroids and assignments are calculated? From the paper I understood that only the proxy noise function is used an then the quantization is finished, or is this wrong?

Second question:
For the first part I wrapped only the second fully connected layer with the `quant_noise()` method (no specific reason just for trying to understand the method). So before each forward, random blocks of the FC layer will get dropped (i.e. zeroed). But when I do this with just a few batches, soon the whole weight matrix has zero values. Isn't this where the STE would be applied? According to your paper, for the noisy weights STE is used and for the other ones the normal gradient is calculated. If so, is this missing here or am I misunderstanding something?:

https://github.com/pytorch/fairseq/blob/148327d8c1e3a5f9d17a11bbb1973a7cf3f955d3/fairseq/modules/quant_noise.py#L100-L104

Technically, in the backward pass the zero weights would not stay zero but be replaced with themselves like they were before the forward pass happened right?

#### Code
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from fairseq.modules.quant_noise import quant_noise


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x


transform = transforms.Compose(
    [transforms.ToTensor(),
     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

trainset = torchvision.datasets.CIFAR10(root='./data-bin', train=True,
                                        download=True, transform=transform)
trainloader = torch.utils.data.DataLoader(trainset, batch_size=8,
                                          shuffle=True, num_workers=0)

testset = torchvision.datasets.CIFAR10(root='./data-bin', train=False,
                                       download=True, transform=transform)
testloader = torch.utils.data.DataLoader(testset, batch_size=8,
                                         shuffle=False, num_workers=0)

classes = ('plane', 'car', 'bird', 'cat',
           'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

net = Net()

quant_noise(net.fc2, p=0.5, block_size=20)

criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

for epoch in range(3):  # loop over the dataset multiple times

    running_loss = 0.0
    for i, data in enumerate(trainloader, 0):

        # get the inputs; data is a list of [inputs, labels]
        inputs, labels = data
        labels = labels.to(torch.float32)
        # zero the parameter gradients
        optimizer.zero_grad()

        # forward + backward + optimize
        outputs = net(inputs)
        loss = criterion(outputs, labels.to(torch.long))
        loss.backward()
        optimizer.step()

        quantized_array = net.fc2.weight.detach().numpy()

        # print statistics
        running_loss += loss.item()
        if i % 2000 == 1999:    # print every 2000 mini-batches
            print('[{:d}, {:5d}] loss: {:.3f}'
                  .format(epoch + 1, i + 1, running_loss / 2000))
            running_loss = 0.0

print("Finished Training")

```

#### What have you tried?
I tested the observation from my question by looking at `quantized_array` in every iteration of the loop (just a few are sufficient with `p=0.5`). Note that for lower values of `p` this goes of course slower but I think it has the same outcome.
I am running this on CPU.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): master
 - PyTorch Version (e.g., 1.0): 1.7.1, `torchvision=0.8.2`
 - OS (e.g., Linux): Linux 5.4.89-1-MANJARO
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): `pip install --editable ./` inside fairseq root folder
 - Python version: 3.8
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information:

**Comment by lematt1991** [comment]
CC @huihuifan

**Comment by huihuifan** [comment]
@pierrestock ?

**Comment by pdeubel** [comment]
So if I am correct and the `quant_noise()` function in `modules.quant_noise` does not use the STE, a simple solution for this would be to register a forward hook (since it is called after the forward computations) on the to be quantized module where the dropped weights from the pre forward hook are reintroduced into the weight matrix. This way these weights use the STE (passed through) and the for the weights that were actually used in the forward their gradient is calculated as usual. (I could make a PR for this if you think this is right)
